agent/lm_agent/backend_utils/utils.py

```diff
@@ -0,0 +1,357 @@
+        raise LicenseManagerBackendConnectionError("Could not connect to the backend health-check endpoint")
+
+
+async def get_jobs_from_backend() -> Optional[List[JobSchema]]:
+    """
+    Get all jobs for the cluster with its bookings from the backend.
+    """
+    cluster_data = await get_cluster_from_backend()
+    if cluster_data:
+        return cluster_data.jobs
+    return []
+
+
+async def get_configs_from_backend() -> List[ConfigurationSchema]:
+    """
+    Get all config rows from the backend.
+    """
+    cluster_data = await get_cluster_from_backend()
+    if cluster_data:
+        return cluster_data.configurations
+    return []
+
+
+async def get_all_clusters_from_backend() -> List[ClusterSchema]:
+    """
+    Get all clusters from the backend.
+    """
+    async with AsyncBackendClient() as backend_client:
+        resp = await backend_client.get("/lm/clusters")
+
+        LicenseManagerBackendConnectionError.require_condition(
+            resp.status_code == 200, f"Could not get cluster data from the backend: {resp.text}"
+        )
+
+    parsed_resp: List = resp.json()
+
+    clusters = []
+
+    for cluster in parsed_resp:
+        try:
+            parsed_cluster = ClusterSchema.parse_obj(cluster)
+        except ValidationError as err:
+            logger.error(f"Could not validate cluster data: {str(err)}")
+            raise LicenseManagerParseError("Could not parse cluster data returned from the backend")
+
+        clusters.append(parsed_cluster)
+
+    return clusters
+
+
+async def get_cluster_from_backend() -> ClusterSchema:
+    """
+    Get cluster data from the backend.
+    """
+    async with AsyncBackendClient() as backend_client:
+        resp = await backend_client.get("/lm/clusters/by_client_id")
+        LicenseManagerBackendConnectionError.require_condition(
+            resp.status_code == 200, f"Could not get cluster data from the backend: {resp.text}"
+        )
+
+    parsed_resp: dict = resp.json()
+
+    try:
+        cluster_data = ClusterSchema.parse_obj(parsed_resp)
+    except ValidationError as err:
+        logger.error(f"Could not validate cluster data: {str(err)}")
+        raise LicenseManagerParseError("Could not parse cluster data returned from the backend")
+
+    return cluster_data
+
+
+def get_feature_ids(cluster_data: ClusterSchema) -> Dict[str, int]:
+    """
+    Get the feature_id for each product_feature in the cluster.
+    """
+    features_id = {
+        f"{feature.product.name}.{feature.name}": feature.id
+        for configuration in cluster_data.configurations
+        for feature in configuration.features
+    }
+
+    return features_id
+
+
+def get_grace_times(cluster_data: ClusterSchema) -> Dict[int, int]:
+    """
+    Get the grace time for each feature_id in the cluster.
+    """
+    grace_times = {
+        feature.id: configuration.grace_time
+        for configuration in cluster_data.configurations
+        for feature in configuration.features
+    }
```

Comment:
Despite I think this code won't bring any performance problems **exclusively** due to the potential size of the database, I have a suggestion that I believe can improve the code base. Here it is the snippet I used to test two different approaches:

```python
import timeit
import itertools

pizzas = [{'pizza': 'margherita'}, {'pizza': 'hawaiian'}, {'pizza': 'pepperoni'}, {'pizza': 'veggie'}, {'pizza': 'meat lovers'}, {'pizza': 'vegan'}, {'pizza': 'supreme'}]
cheeses = [{'cheese': 'mozzarella'}, {'cheese': 'cheddar'}, {'cheese': 'parmesan'}]

def build_variants_with_compreehension():
    variants = {
        pizza.get("pizza"): cheese.get("cheese")
        for cheese in cheeses
        for pizza in pizzas
    }
    return variants

def build_variants_with_map():
    variants = dict(map(lambda x: (x[0]['pizza'], x[1]['cheese']), itertools.product(pizzas, cheeses)))
    return variants

time_to_run_comprehension = timeit.timeit(build_variants_with_compreehension, number=100000)
time_to_run_map = timeit.timeit(build_variants_with_map, number=100000)

print(f"Using map is {time_to_run_map/time_to_run_comprehension} times faster than comprehension")
```

Reply:
I feel that the improvement in the execution time is not worth decreasing the readability.

Reply:
don't you think that `dict(map(lambda x: (x[0]['pizza'], x[1]['cheese']), itertools.product(pizzas, cheeses)))` is readable at all?

Reply:
Using comprehensions is generally best practice in python. We should prefer this approach unless the performance gain is very substatial.